jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp

```diff
@@ -361,3 +497,58 @@ void convert_threadblock(symbol_tablet &symbol_table)
     }
   }
 }
+
+/// Iterate through the symbol table to find and appropriately instrument
+/// synchronized methods.
+///
+/// A synchronized method is a method with the synchronization keyword. Any
+/// code inside of such methods has to be thread-safe. To this end, a call to
+/// "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V" and
+/// "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V" is instrumented
+/// at the start and end of the method. Note, that the former is instrumented
+/// at every statement where the execution can exit the method in question.
+/// Specifically, out of order return statements and exceptions. The latter
+/// is dealt with by instrumenting a try catch block.
+///
+/// Note: current version cannot handle static methods.
+///
+/// Note': This method requires the availability of
+///   "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V" and
+///   "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V".
+///   (java-library-models)
+///
+/// \param symbol_table: a symbol table
+void convert_synchronized_methods(symbol_tablet &symbol_table)
+{
+  for(const auto &entry : symbol_table)
+  {
+    const symbolt &symbol = entry.second;
+
+    if(symbol.type.id() != ID_code)
+      continue;
+    if(symbol.value.is_nil())
+      continue;
+    if(symbol.type.get(ID_is_synchronized) != "1")
+      continue;
+
+    bool is_static = (symbol.type.get(ID_is_static) == "1");
+    if(is_static)
+      // FIXME: handle static synchronized methods
```

Comment:
Ok, static synchronized methods seem to be treated as simply static methods. This will produce many dataraces and unexpected results. Warn the user about this. Example: https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/remove_function.cpp#L44

Reply:
yes as the synchronization semantics for static methods is  different (lock on the class instead the object). This is limitation with the current implementation. I have no issues with adding the warning  method, but I think it makes more sense to add it outside (where we  set the static/synchronized  irep flags)  such that we don't have to pass around the message handler reference.

Reply:
Yes, makes sense.
In that case use a `PRECONDITION` in this method to ensure it won't be called with a static synchronized method, and do the warning outside.

Reply:
done